hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -248,4 +368,46 @@ private Address asMirrorAddress(final Address addressOrAlias, final MessageFrame
         final var aliases = ((HederaStackedWorldStateUpdater) messageFrame.getWorldUpdater()).aliases();
         return aliases.resolveForEvm(addressOrAlias);
     }
+
+    private Optional<SolidityAction> popActionStack(final MessageFrame frame) {
+        if (!currentActionsStack.isEmpty()) {
+            return Optional.of(currentActionsStack.pop());
+        } else {
+            log.atDebug().log(
+                    "HederaTracer action stack prematurely empty ({})", () -> formatFrameContextForLog(frame));
+            return Optional.empty();
```

Comment:
This appears to indicate things are very broken (is it even impossible to inducde in tests?), so a log at debug seems too low.  I wounder if the stack trace from poping an empty stack would be more usefule for devops, because it would result in a TX level exception that will receive scary logs.  I know sonar is bugging you about the exeption but I think we want it, or at least log it at error with a stack trace.

Reply:
I wondered about that level myself, but thought that this followed the recent logging policy announcement.  Those guidelines say not to log anything at `info` or above that is going to possibly spew, possibly for a lot of transactions.  And I suppose that is possible here: some app, being tested, may trigger this over and over.  (Another solution would be a higher level + a throttle.  I'll look into that.)

Reply:
(w.r.t. include in tests: it is of course testable in _unit_ tests and I _do_ test it there.  I don't think I can drive it to this state in the integration tests though.  At least, I hope not.)

Reply:
Log level should be warn.  If it spews it is horribly broken and needs to be turned off.  Based on the construction of the tracer the number of logs should be zero if we have the logic correct.  So normal operation is zero logs, and it is a recoverable error, hence warn not error.